### PR TITLE
Upgrade `temp` package to 0.9 (Add Promise return types for async functions)

### PR DESCRIPTION
**Pull request by abettadapur**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/bruce/node-temp/commit/e649841409ec2d0421cdb39f564f79849518f99e>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by abettadapur** [comment]
@DanielRosenwasser who should I ask for a review on this?

File: types/temp/index.d.ts [hunk 1]
```diff
@@ -1,48 +1,51 @@
...
 
-  let dir: string;
+    let dir: string;
 
-  function track(value?: boolean): typeof temp;
+    function track(value?: boolean): typeof temp;
 
-  function mkdir(affixes?: string | AffixOptions, callback?: (err: any, dirPath: string) => void): void;
+    function mkdir(affixes?: string | AffixOptions): Promise<string>;
+    function mkdir(affixes?: string | AffixOptions, callback?: (err: any, dirPath: string) => void): void;
```
Comment (amcasey) [inline types/temp/index.d.ts:32]: Looks like `callback` is actually non-optional in this overload then?

**Review by amcasey — APPROVED** [review]
Optional suggestion.

**Comment by DanielRosenwasser** [comment]
@abettadapur sorry for the delay. I actually don't use this package much anymore, so if you'd be open to it, you can add yourself as a maintainer by adding yourself to the file header.

**Pull request closed without merging**